File: src/flavor-rs/src/psp/format_2025/policy.rs
// SPDX-FileCopyrightText: Copyright (c) 2025 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Launch-time policy enforcement for the Rust launcher.

use std::fs;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

use crate::exceptions::{FlavorError, Result};

/// Package-declared constraints (from package metadata).
#[derive(Default, Debug, serde::Deserialize)]
pub struct PackagePolicy {
    pub platforms: Vec<String>,
    pub refuse_root: bool,
    pub max_age_days: Option<u64>,
    pub require_env: Vec<String>,
}

/// Operator policy (from policy.toml).
#[derive(Default, Debug)]
pub struct OperatorPolicy {
    pub require_trusted_key: bool,
    pub use_os_keychain: bool,
    pub refuse_root: bool,
    pub max_age_days: Option<u64>,
    pub allow_platforms: Vec<String>,
    pub require_sbom: bool,
}

/// Merged policy: stricter wins.
#[derive(Default, Debug)]
pub struct EffectivePolicy {
    pub platforms: Vec<String>,
    pub refuse_root: bool,
    pub max_age_days: Option<u64>,
    pub require_env: Vec<String>,
    pub require_trusted_key: bool,
    pub use_os_keychain: bool,
    pub require_sbom: bool,
}

#[derive(Default, Debug, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct OperatorPolicyFile {
    #[serde(default)]
    trust: Option<TrustSection>,
    #[serde(default)]
    execution: Option<ExecutionSection>,
    #[serde(default)]
    attestation: Option<AttestationSection>,
}

#[derive(Default, Debug, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct TrustSection {
    #[serde(default)]
    require_trusted_key: Option<bool>,
    #[serde(default)]
    use_os_keychain: Option<bool>,
}

#[derive(Default, Debug, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct ExecutionSection {
    #[serde(default)]
    refuse_root: Option<bool>,
    #[serde(default)]
    max_age_days: Option<u64>,
    #[serde(default)]
    allow_platforms: Option<Vec<String>>,
}

#[derive(Default, Debug, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct AttestationSection {
    #[serde(default)]
    require_sbom: Option<bool>,
}

fn get_system_policy_path() -> PathBuf {
    #[cfg(target_os = "windows")]
    {
        if let Ok(pd) = std::env::var("PROGRAMDATA") {
            return PathBuf::from(pd).join("flavor").join("policy.toml");
        }
        return PathBuf::from("C:\\ProgramData\\flavor\\policy.toml");
    }
    #[cfg(not(target_os = "windows"))]
    PathBuf::from("/etc/flavor/policy.toml")
}

fn get_user_policy_path() -> Option<PathBuf> {
    if let Ok(dir) = std::env::var(crate::env_vars::CONFIG_DIR) {
        return Some(PathBuf::from(dir).join("policy.toml"));
    }
    if let Ok(xdg) = std::env::var("XDG_CONFIG_HOME") {
        return Some(PathBuf::from(xdg).join("flavor").join("policy.toml"));
    }
    #[cfg(not(target_os = "windows"))]
    if let Ok(home) = std::env::var("HOME") {
        return Some(
            PathBuf::from(home)
                .join(".config")
                .join("flavor")
                .join("policy.toml"),
        );
    }
    None
}

fn apply_operator_policy_file(policy: &mut OperatorPolicy, file: OperatorPolicyFile) {
    if let Some(trust) = file.trust {
        if let Some(value) = trust.require_trusted_key {
            policy.require_trusted_key = value;
        }
        if let Some(value) = trust.use_os_keychain {
            policy.use_os_keychain = value;
        }
    }
    if let Some(execution) = file.execution {
        if let Some(value) = execution.refuse_root {
            policy.refuse_root = value;
        }
        if let Some(value) = execution.max_age_days {
            policy.max_age_days = Some(value);
        }
        if let Some(value) = execution.allow_platforms {
            policy.allow_platforms = value;
        }
    }
    if let Some(attestation) = file.attestation
        && let Some(value) = attestation.require_sbom
    {
        policy.require_sbom = value;
    }
}

fn parse_policy_file(content: &str, policy: &mut OperatorPolicy) -> Result<()> {
    let file: OperatorPolicyFile = toml::from_str(content)
        .map_err(|e| FlavorError::Generic(format!("invalid policy.toml: {e}")))?;
    apply_operator_policy_file(policy, file);
    Ok(())
}

/// Load operator policy from system and user files.
pub fn load_operator_policy() -> Result<OperatorPolicy> {
    let system_path = get_system_policy_path();
    let user_path = get_user_policy_path();
    load_operator_policy_from_paths(Some(&system_path), user_path.as_deref())
}

fn load_operator_policy_from_paths(
    system_path: Option<&Path>,
    user_path: Option<&Path>,
) -> Result<OperatorPolicy> {
    let mut policy = OperatorPolicy::default();
    if let Some(path) = system_path {
        if let Some(content) = read_policy_file(path)? {
            parse_policy_file(&content, &mut policy)?;
        }
    }
    if let Some(path) = user_path {
        if let Some(content) = read_policy_file(path)? {
            parse_policy_file(&content, &mut policy)?;
        }
    }
    Ok(policy)
}

fn read_policy_file(path: &Path) -> Result<Option<String>> {
    match fs::read_to_string(path) {
        Ok(content) => Ok(Some(content)),
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(err) => Err(FlavorError::Generic(format!(
            "reading policy {}: {}",
            path.display(),
            err
        ))),
    }
}

/// Merge package + operator policy. Stricter always wins.
pub fn merge_policy(pkg: PackagePolicy, op: OperatorPolicy) -> EffectivePolicy {
    let platforms = if !pkg.platforms.is_empty() && !op.allow_platforms.is_empty() {
        pkg.platforms
            .iter()
            .filter(|p| op.allow_platforms.contains(p))
            .cloned()
            .collect()
    } else if !op.allow_platforms.is_empty() {
        op.allow_platforms.clone()
    } else {
        pkg.platforms.clone()
    };

    let refuse_root = pkg.refuse_root || op.refuse_root;

    let max_age_days = match (pkg.max_age_days, op.max_age_days) {
        (Some(a), Some(b)) => Some(a.min(b)),
        (Some(a), None) => Some(a),
        (None, b) => b,
    };

    EffectivePolicy {
        platforms,
        refuse_root,
        max_age_days,
        require_env: pkg.require_env,
        require_trusted_key: op.require_trusted_key,
        use_os_keychain: op.use_os_keychain,
        require_sbom: op.require_sbom,
    }
}

/// Check whether the current process is running with Windows Administrator privileges.
/// Uses `CheckTokenMembership` against the built-in Administrators SID (S-1-5-32-544).
/// Passing `None` as the token handle checks the effective token of the current process.
#[cfg(target_os = "windows")]
#[allow(unsafe_code)]
fn is_windows_admin() -> bool {
    use windows::Win32::Foundation::BOOL;
    use windows::Win32::Security::{
        AllocateAndInitializeSid, CheckTokenMembership, FreeSid, PSID, SID_IDENTIFIER_AUTHORITY,
    };
    use windows::Win32::System::SystemServices::{
        DOMAIN_ALIAS_RID_ADMINS, SECURITY_BUILTIN_DOMAIN_RID,
    };

    const NT_AUTHORITY: SID_IDENTIFIER_AUTHORITY = SID_IDENTIFIER_AUTHORITY {
        Value: [0, 0, 0, 0, 0, 5],
    };

    unsafe {
        let mut admin_sid = PSID::default();
        if AllocateAndInitializeSid(
            &NT_AUTHORITY,
            2,
            SECURITY_BUILTIN_DOMAIN_RID as u32,
            DOMAIN_ALIAS_RID_ADMINS as u32,
            0,
            0,
            0,
            0,
            0,
            0,
            &mut admin_sid,
        )
        .is_err()
        {
            return false;
        }

        let mut is_member = BOOL::default();
        let ok = CheckTokenMembership(None, admin_sid, &mut is_member);
        let _ = FreeSid(admin_sid);
        ok.is_ok() && is_member.as_bool()
    }
}

/// Enforce policy against current runtime environment.
/// Returns Err with a descriptive message on first violation.
/// `key_trusted` is false only when the trusted store exists AND the key is explicitly absent.
#[allow(unsafe_code)] // Required for libc::geteuid() FFI call
pub fn enforce_policy(
    policy: &EffectivePolicy,
    build_timestamp: u64,
    has_sbom: bool,
    key_trusted: bool,
) -> std::result::Result<(), String> {
    let current_platform = get_current_platform();

    // 1. Platform check
    if !policy.platforms.is_empty() && !policy.platforms.contains(&current_platform) {
        return Err(format!(
            "platform not permitted: {} not in {:?}",
            current_platform, policy.platforms
        ));
    }

    // 2. OS keychain check
    if policy.use_os_keychain {
        return Err("use_os_keychain is not supported by this launcher".to_string());
    }

    // 3. Root / Administrator check
    #[cfg(unix)]
    if policy.refuse_root && unsafe { libc::geteuid() } == 0 {
        return Err("refused to run as root or Administrator".to_string());
    }
    #[cfg(target_os = "windows")]
    if policy.refuse_root && is_windows_admin() {
        return Err("refused to run as root or Administrator".to_string());
    }

    // 4. Age check
    if let Some(max_days) = policy.max_age_days {
        if build_timestamp > 0 {
            let now = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs();
            let age_days = now.saturating_sub(build_timestamp) / 86400;
            if age_days > max_days {
                return Err(format!(
                    "package is {} days old — policy requires max {} days",
                    age_days, max_days
                ));
            }
        }
    }

    // 5. Environment variable check
    for var in &policy.require_env {
        match std::env::var(var) {
            Ok(val) if !val.is_empty() => {}
            _ => return Err(format!("required environment variable not set: {}", var)),
        }
    }

    // 6. SBOM check
    if policy.require_sbom && !has_sbom {
        return Err(
            "package built without attestation slot — operator policy requires SBOM".to_string(),
        );
    }

    // 7. Trusted key check
    if policy.require_trusted_key && !key_trusted {
        return Err(
            "operator policy requires a trusted signing key — package key is not in the trusted store"
                .to_string(),
        );
    }

    Ok(())
}

pub fn get_current_platform() -> String {
    let os = if cfg!(target_os = "linux") {
        "linux"
    } else if cfg!(target_os = "macos") {
        "darwin"
    } else if cfg!(target_os = "freebsd") {
        "freebsd"
    } else if cfg!(target_os = "windows") {
        "windows"
    } else {
        std::env::consts::OS
    };
    let arch = if cfg!(target_arch = "aarch64") {
        "arm64"
    } else {
        "amd64"
    };
    format!("{}_{}", os, arch)
}

#[cfg(test)]
#[allow(unsafe_code)] // Required for env::set_var/remove_var (unsafe in Rust edition 2024)
mod tests {
    use super::*;
    use std::env;

    #[test]
    fn test_get_current_platform_nonempty() {
        let p = get_current_platform();
        assert!(!p.is_empty());
        assert!(p.contains('_'));
    }

    #[test]
    fn test_merge_policy_refuse_root() {
        let pkg = PackagePolicy {
            refuse_root: false,
            ..Default::default()
        };
        let op = OperatorPolicy {
            refuse_root: true,
            ..Default::default()
        };
        let eff = merge_policy(pkg, op);
        assert!(eff.refuse_root);
    }

    #[test]
    fn test_merge_policy_max_age_days_min() {
        let pkg = PackagePolicy {
            max_age_days: Some(365),
            ..Default::default()
        };
        let op = OperatorPolicy {
            max_age_days: Some(90),
            ..Default::default()
        };
        let eff = merge_policy(pkg, op);
        assert_eq!(eff.max_age_days, Some(90));
    }

    #[test]
    fn test_merge_policy_max_age_only_pkg() {
        let pkg = PackagePolicy {
            max_age_days: Some(180),
            ..Default::default()
        };
        let op = OperatorPolicy::default();
        let eff = merge_policy(pkg, op);
        assert_eq!(eff.max_age_days, Some(180));
    }

    #[test]
    fn test_merge_policy_platforms_intersection() {
        let pkg = PackagePolicy {
            platforms: vec!["linux_amd64".to_string(), "darwin_arm64".to_string()],
            ..Default::default()
        };
        let op = OperatorPolicy {
            allow_platforms: vec!["linux_amd64".to_string(), "linux_arm64".to_string()],
            ..Default::default()
        };
        let eff = merge_policy(pkg, op);
        assert_eq!(eff.platforms, vec!["linux_amd64".to_string()]);
    }

    #[test]
    fn test_merge_policy_platforms_only_operator() {
        let pkg = PackagePolicy::default();
        let op = OperatorPolicy {
            allow_platforms: vec!["linux_amd64".to_string()],
            ..Default::default()
        };
        let eff = merge_policy(pkg, op);
        assert_eq!(eff.platforms, vec!["linux_amd64".to_string()]);
    }

    #[test]
    fn test_merge_policy_propagates_os_keychain_flag() {
        let pkg = PackagePolicy::default();
        let op = OperatorPolicy {
            use_os_keychain: true,
            ..Default::default()
        };
        let eff = merge_policy(pkg, op);
        assert!(eff.use_os_keychain);
    }

    #[test]
    fn test_enforce_policy_permissive() {
        let eff = EffectivePolicy::default();
        assert!(enforce_policy(&eff, 0, false, true).is_ok());
    }

    #[test]
    fn test_enforce_policy_platform_blocked() {
        let eff = EffectivePolicy {
            platforms: vec!["__nonexistent__".to_string()],
            ..Default::default()
        };
        assert!(enforce_policy(&eff, 0, false, true).is_err());
    }

    #[test]
    fn test_enforce_policy_sbom_required() {
        let eff = EffectivePolicy {
            require_sbom: true,
            ..Default::default()
        };
        assert!(enforce_policy(&eff, 0, false, true).is_err());
        assert!(enforce_policy(&eff, 0, true, true).is_ok());
    }

    #[test]
    fn test_enforce_policy_env_var_missing() {
        let eff = EffectivePolicy {
            require_env: vec!["__FLAVOR_POLICY_TEST_NONEXISTENT__".to_string()],
            ..Default::default()
        };
        // Make sure the variable is unset
        unsafe { env::remove_var("__FLAVOR_POLICY_TEST_NONEXISTENT__") };
        assert!(enforce_policy(&eff, 0, false, true).is_err());
    }

    #[test]
    fn test_enforce_policy_env_var_present() {
        unsafe { env::set_var("__FLAVOR_POLICY_TEST__", "1") };
        let eff = EffectivePolicy {
            require_env: vec!["__FLAVOR_POLICY_TEST__".to_string()],
            ..Default::default()
        };
        assert!(enforce_policy(&eff, 0, false, true).is_ok());
        unsafe { env::remove_var("__FLAVOR_POLICY_TEST__") };
    }

    #[test]
    fn test_enforce_policy_age_exceeded() {
        let eff = EffectivePolicy {
            max_age_days: Some(0),
            ..Default::default()
        };
        // Build timestamp of 1 (ancient) should exceed 0 day limit
        assert!(enforce_policy(&eff, 1, false, true).is_err());
    }

    #[test]
    fn test_enforce_policy_require_trusted_key_untrusted() {
        let eff = EffectivePolicy {
            require_trusted_key: true,
            ..Default::default()
        };
        assert!(enforce_policy(&eff, 0, false, false).is_err());
    }

    #[test]
    fn test_enforce_policy_require_trusted_key_trusted() {
        let eff = EffectivePolicy {
            require_trusted_key: true,
            ..Default::default()
        };
        assert!(enforce_policy(&eff, 0, false, true).is_ok());
    }

    #[test]
    fn test_enforce_policy_require_trusted_key_not_required() {
        let eff = EffectivePolicy {
            require_trusted_key: false,
            ..Default::default()
        };
        // Even untrusted key should pass when policy doesn't require it
        assert!(enforce_policy(&eff, 0, false, false).is_ok());
    }

    #[test]
    fn test_load_operator_policy_missing_file() {
        let system = std::path::PathBuf::from("/tmp/__nonexistent_flavor_system_policy.toml");
        let user = std::path::PathBuf::from("/tmp/__nonexistent_flavor_user_policy.toml");
        let policy =
            load_operator_policy_from_paths(Some(&system), Some(&user)).expect("policy load");
        assert!(!policy.require_trusted_key);
    }

    #[test]
    fn test_load_operator_policy_invalid_system_file_errors() {
        let system_dir = tempfile::TempDir::new().expect("tempdir");
        let system_file = system_dir.path().join("policy.toml");
        std::fs::write(&system_file, b"[trust\nrequire_trusted_key = true\n")
            .expect("write bad policy");
        let result = load_operator_policy_from_paths(Some(&system_file), None);
        assert!(result.is_err(), "invalid system policy must fail closed");
    }

    #[test]
    fn test_load_operator_policy_invalid_user_file_errors() {
        let user_dir = tempfile::TempDir::new().expect("tempdir");
        let user_file = user_dir.path().join("policy.toml");
        std::fs::write(
            &user_file,
            b"[trust]\nrequire_trusted_key = true\nunknown_key = true\n",
        )
        .expect("write bad policy");
        let result = load_operator_policy_from_paths(None, Some(&user_file));
        assert!(result.is_err(), "invalid user policy must fail closed");
    }

    #[test]
    fn test_parse_policy_file_all_fields() {
        let content = "[trust]\nrequire_trusted_key = true\nuse_os_keychain = true\n[execution]\nrefuse_root = true\nmax_age_days = 90\n[attestation]\nrequire_sbom = true\n";
        let mut policy = OperatorPolicy::default();
        parse_policy_file(content, &mut policy).expect("parse policy");
        assert!(policy.require_trusted_key);
        assert!(policy.use_os_keychain);
        assert!(policy.refuse_root);
        assert_eq!(policy.max_age_days, Some(90));
        assert!(policy.require_sbom);
    }

    #[test]
    fn test_parse_policy_file_ignores_comments() {
        let content = "# top comment\n[trust]\n# inline comment\nrequire_trusted_key = true\n";
        let mut policy = OperatorPolicy::default();
        parse_policy_file(content, &mut policy).expect("parse policy");
        assert!(policy.require_trusted_key);
    }

    #[test]
    fn test_parse_policy_file_inline_comment() {
        let content = "[execution]\nrefuse_root = true # disable if needed\n";
        let mut policy = OperatorPolicy::default();
        parse_policy_file(content, &mut policy).expect("parse policy");
        assert!(policy.refuse_root, "should strip inline comment");
    }

    #[test]
    fn test_parse_policy_file_allow_platforms() {
        let content = "[execution]\nallow_platforms = [\"linux_amd64\", \"linux_arm64\"]\n";
        let mut policy = OperatorPolicy::default();
        parse_policy_file(content, &mut policy).expect("parse policy");
        assert_eq!(policy.allow_platforms, vec!["linux_amd64", "linux_arm64"]);
    }

    #[test]
    fn test_parse_policy_file_preserves_hash_inside_string() {
        let content = "[execution]\nallow_platforms = [\"linux_amd64#beta\"]\n";
        let mut policy = OperatorPolicy::default();
        parse_policy_file(content, &mut policy).expect("parse policy");
        assert_eq!(policy.allow_platforms, vec!["linux_amd64#beta"]);
    }

    #[test]
    fn test_parse_policy_file_rejects_unknown_keys() {
        let content = "[trust]\nrequire_trusted_key = true\nunknown_key = true\n";
        let mut policy = OperatorPolicy::default();
        let err = parse_policy_file(content, &mut policy).expect_err("unknown key must fail");
        let err = err.to_string();
        assert!(err.contains("unknown_key") || err.contains("unknown"));
    }

    #[test]
    fn test_enforce_policy_rejects_use_os_keychain() {
        let eff = EffectivePolicy {
            use_os_keychain: true,
            ..Default::default()
        };
        assert!(enforce_policy(&eff, 0, false, true).is_err());
    }

    #[test]
    fn test_enforce_policy_env_var_empty_string_is_absent() {
        unsafe { env::set_var("__FLAVOR_POLICY_EMPTY__", "") };
        let eff = EffectivePolicy {
            require_env: vec!["__FLAVOR_POLICY_EMPTY__".to_string()],
            ..Default::default()
        };
        assert!(enforce_policy(&eff, 0, false, true).is_err());
        unsafe { env::remove_var("__FLAVOR_POLICY_EMPTY__") };
    }

    #[test]
    fn test_merge_then_enforce_empty_intersection_unrestricted() {
        // Disjoint platform sets → intersection=[] → enforce_policy treats [] as unrestricted
        let pkg = PackagePolicy {
            platforms: vec!["linux_amd64".to_string()],
            ..Default::default()
        };
        let op = OperatorPolicy {
            allow_platforms: vec!["darwin_arm64".to_string()],
            ..Default::default()
        };
        let eff = merge_policy(pkg, op);
        assert!(eff.platforms.is_empty());
        assert!(enforce_policy(&eff, 0, false, true).is_ok());
    }
}

File: src/flavor/config/policy.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025 provide.io llc. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#

"""FlavorPack launch-time policy: schema, parsing, and merge logic."""

from __future__ import annotations

from datetime import UTC, datetime
from enum import StrEnum
import json
import os
from pathlib import Path
import platform
import sys
from typing import Any

from attrs import define, field

from flavor.config.dirs import get_policy_file
from flavor.console import get_command_logger

log = get_command_logger("config.policy")

POLICY_VERSION = 1

_POLICY_SCHEMA: dict[str, dict[str, object]] = {
    "trust": {
        "require_trusted_key": bool,
        "use_os_keychain": bool,
    },
    "execution": {
        "refuse_root": bool,
        "max_age_days": int,
        "allow_platforms": list,
    },
    "attestation": {
        "require_sbom": bool,
    },
    "enforcement": {
        "default": str,
        "platform_mismatch": str,
        "untrusted_key": str,
        "expired_package": str,
        "missing_env": str,
        "missing_sbom": str,
        "root_execution": str,
        "os_keychain": str,
    },
}

_VALID_ENFORCEMENT_MODES = {"deny", "warn", "allow"}


class EnforcementMode(StrEnum):
    """How a policy check violation is handled."""

    DENY = "deny"
    WARN = "warn"
    ALLOW = "allow"


@define
class EnforcementPolicy:
    """Per-check enforcement modes. Omitted checks inherit from default."""

    default: EnforcementMode = field(default=EnforcementMode.DENY)
    platform_mismatch: EnforcementMode | None = field(default=None)
    root_execution: EnforcementMode | None = field(default=None)
    expired_package: EnforcementMode | None = field(default=None)
    missing_env: EnforcementMode | None = field(default=None)
    missing_sbom: EnforcementMode | None = field(default=None)
    untrusted_key: EnforcementMode | None = field(default=None)
    os_keychain: EnforcementMode | None = field(default=None)

    def mode_for(self, check: str) -> EnforcementMode:
        """Return the enforcement mode for a given check name."""
        val = getattr(self, check, None)
        return val if val is not None else self.default


@define
class PackagePolicy:
    """Constraints declared by the package builder in pyproject.toml."""

    platforms: list[str] = field(factory=list)
    refuse_root: bool = field(default=False)
    max_age_days: int | None = field(default=None)
    require_env: list[str] = field(factory=list)


@define
class OperatorPolicy:
    """Operator overlay from policy.json."""

    require_trusted_key: bool = field(default=False)
    use_os_keychain: bool = field(default=False)
    refuse_root: bool = field(default=False)
    max_age_days: int | None = field(default=None)
    allow_platforms: list[str] = field(factory=list)
    require_sbom: bool = field(default=False)
    enforcement: EnforcementPolicy = field(factory=EnforcementPolicy)


@define
class EffectivePolicy:
    """Merged policy: the stricter of package + operator wins per field."""

    platforms: list[str] = field(factory=list)
    refuse_root: bool = field(default=False)
    max_age_days: int | None = field(default=None)
    require_env: list[str] = field(factory=list)
    require_trusted_key: bool = field(default=False)
    use_os_keychain: bool = field(default=False)
    require_sbom: bool = field(default=False)
    enforcement: EnforcementPolicy = field(factory=EnforcementPolicy)


def parse_package_policy(raw: dict[str, Any]) -> PackagePolicy:
    """Parse [tool.flavor.policy] dict from pyproject.toml into a PackagePolicy."""
    return PackagePolicy(
        platforms=raw.get("platforms", []),
        refuse_root=bool(raw.get("refuse_root", False)),
        max_age_days=raw.get("max_age_days"),
        require_env=raw.get("require_env", []),
    )


def _parse_enforcement_section(raw: dict[str, Any]) -> EnforcementPolicy:
    """Parse the enforcement section of a policy file."""
    kwargs: dict[str, Any] = {}
    for key, value in raw.items():
        if key not in _POLICY_SCHEMA["enforcement"]:
            raise ValueError(f"unknown enforcement key {key!r}")
        if value not in _VALID_ENFORCEMENT_MODES:
            raise ValueError(f"enforcement.{key} must be one of {_VALID_ENFORCEMENT_MODES}, got {value!r}")
        kwargs[key] = EnforcementMode(value)
    return EnforcementPolicy(**kwargs)


def _parse_operator_policy(raw: dict[str, Any]) -> OperatorPolicy:
    """Parse policy file content into an OperatorPolicy."""
    trust = raw.get("trust", {})
    execution = raw.get("execution", {})
    attestation = raw.get("attestation", {})
    enforcement_raw = raw.get("enforcement", {})

    enforcement = _parse_enforcement_section(enforcement_raw) if enforcement_raw else EnforcementPolicy()

    return OperatorPolicy(
        require_trusted_key=bool(trust.get("require_trusted_key", False)),
        use_os_keychain=bool(trust.get("use_os_keychain", False)),
        refuse_root=bool(execution.get("refuse_root", False)),
        max_age_days=execution.get("max_age_days"),
        allow_platforms=execution.get("allow_platforms", []),
        require_sbom=bool(attestation.get("require_sbom", False)),
        enforcement=enforcement,
    )


def _raise_invalid_policy(path: Path, message: str) -> None:
    raise ValueError(f"{path}: {message}")


def _validate_operator_policy_value(path: Path, section: str, key: str, value: Any) -> None:
    expected = _POLICY_SCHEMA[section][key]
    label = f"[{section}].{key}"

    if expected is bool:
        if type(value) is not bool:
            _raise_invalid_policy(path, f"{label} must be a boolean")
        return

    if expected is int:
        if type(value) is not int:
            _raise_invalid_policy(path, f"{label} must be an integer")
        return

    if expected is str:
        if type(value) is not str:
            _raise_invalid_policy(path, f"{label} must be a string")
        return

    if expected is list:
        if not isinstance(value, list) or any(type(item) is not str for item in value):
            _raise_invalid_policy(path, f"{label} must be a list of strings")
        return

    _raise_invalid_policy(path, f"{label} has unsupported schema type")


def _validate_operator_policy_file(path: Path, raw: dict[str, Any]) -> None:
    for section, values in raw.items():
        if section == "version":
            if type(values) is not int:
                _raise_invalid_policy(path, "version must be an integer")
            continue
        if section not in _POLICY_SCHEMA:
            _raise_invalid_policy(path, f"unknown policy section [{section}]")
        if not isinstance(values, dict):
            _raise_invalid_policy(path, f"unknown top-level key {section!r}")

        allowed_keys = _POLICY_SCHEMA[section]
        for key, value in values.items():
            if key not in allowed_keys:
                _raise_invalid_policy(path, f"unknown policy key [{section}].{key}")
            _validate_operator_policy_value(path, section, key, value)


def _load_policy_file(path: Path) -> dict[str, Any]:
    try:
        with path.open("r") as f:
            raw: dict[str, Any] = json.load(f)
    except Exception as exc:
        raise ValueError(f"{path}: invalid policy file ({exc})") from exc

    if not isinstance(raw, dict):
        _raise_invalid_policy(path, "policy file root must be an object")

    version = raw.get("version")
    if version is None:
        _raise_invalid_policy(path, "missing required 'version' field")
    if type(version) is not int:
        _raise_invalid_policy(path, "version must be an integer")
    if version > POLICY_VERSION:
        log.warning(
            f"⚠️  {path}: policy version {version} is newer than supported version {POLICY_VERSION} "
            "— unknown fields will be ignored"
        )

    _validate_operator_policy_file(path, raw)
    return raw


def load_operator_policy(*, system: bool = True, user: bool = True) -> OperatorPolicy:
    """Load the operator policy file(s).

    System policy is loaded first, then user policy overrides it.
    If neither file exists, returns a permissive default.
    """
    merged: dict[str, Any] = {}

    if system:
        system_file = get_policy_file(system=True)
        if system_file.exists():
            merged.update(_load_policy_file(system_file))

    if user:
        user_file = get_policy_file(system=False)
        if user_file.exists():
            user_raw = _load_policy_file(user_file)
            for section, values in user_raw.items():
                if section == "version":
                    continue
                if isinstance(values, dict):
                    merged.setdefault(section, {}).update(values)
                else:
                    merged[section] = values

    return _parse_operator_policy(merged)


def merge_policy(pkg: PackagePolicy, op: OperatorPolicy) -> EffectivePolicy:
    """Merge package-declared and operator policies. Stricter always wins."""
    # Platforms: intersection of both non-empty lists
    if pkg.platforms and op.allow_platforms:
        platforms = [p for p in pkg.platforms if p in op.allow_platforms]
    elif op.allow_platforms:
        platforms = list(op.allow_platforms)
    else:
        platforms = list(pkg.platforms)

    refuse_root = pkg.refuse_root or op.refuse_root

    if pkg.max_age_days is not None and op.max_age_days is not None:
        max_age_days: int | None = min(pkg.max_age_days, op.max_age_days)
    elif pkg.max_age_days is not None:
        max_age_days = pkg.max_age_days
    else:
        max_age_days = op.max_age_days

    require_env = list(set(pkg.require_env))

    return EffectivePolicy(
        platforms=platforms,
        refuse_root=refuse_root,
        max_age_days=max_age_days,
        require_env=require_env,
        require_trusted_key=op.require_trusted_key,
        use_os_keychain=op.use_os_keychain,
        require_sbom=op.require_sbom,
        enforcement=op.enforcement,
    )


def get_current_platform() -> str:
    """Return the normalized FlavorPack platform string for the current host."""
    if sys.platform.startswith("linux"):
        os_name = "linux"
    elif sys.platform == "darwin":
        os_name = "darwin"
    elif sys.platform.startswith("freebsd"):
        os_name = "freebsd"
    elif sys.platform == "win32":
        os_name = "windows"
    else:
        os_name = sys.platform
    machine = platform.machine().lower()
    arch = "arm64" if machine in ("arm64", "aarch64") else "amd64"
    return f"{os_name}_{arch}"


def is_privileged_user() -> bool:
    """Return True when the current process is privileged/root."""
    try:
        return os.geteuid() == 0  # type: ignore[attr-defined,unused-ignore]
    except AttributeError:
        return False


def _apply_enforcement(
    mode: EnforcementMode,
    message: str,
    warnings: list[str],
) -> None:
    """Apply enforcement mode: deny raises, warn appends, allow is silent."""
    if mode == EnforcementMode.DENY:
        raise ValueError(message)
    if mode == EnforcementMode.WARN:
        log.warning(f"⚠️  policy warning: {message}")
        warnings.append(message)


def enforce_policy(
    policy: EffectivePolicy,
    build_timestamp: int,
    has_sbom: bool,
    key_trusted: bool,
) -> list[str]:
    """Enforce the effective launch policy for the current runtime environment.

    Returns a list of warning messages for checks in 'warn' mode.
    Raises ValueError for checks in 'deny' mode.
    Checks in 'allow' mode are silently skipped.
    """
    warnings: list[str] = []
    enf = policy.enforcement
    current_platform = get_current_platform()

    # 1. Platform check
    if policy.platforms and current_platform not in policy.platforms:
        _apply_enforcement(
            enf.mode_for("platform_mismatch"),
            f"platform not permitted: {current_platform} not in {policy.platforms}",
            warnings,
        )

    # 2. OS keychain check
    if policy.use_os_keychain:
        _apply_enforcement(
            enf.mode_for("os_keychain"),
            "use_os_keychain is not supported by this launcher",
            warnings,
        )

    # 3. Root / Administrator check
    if policy.refuse_root and is_privileged_user():
        _apply_enforcement(
            enf.mode_for("root_execution"),
            "refused to run as root or Administrator",
            warnings,
        )

    # 4. Age check
    if policy.max_age_days is not None and build_timestamp > 0:
        age_days = int((datetime.now(UTC).timestamp() - build_timestamp) / 86400)
        if age_days > policy.max_age_days:
            _apply_enforcement(
                enf.mode_for("expired_package"),
                f"package is {age_days} days old — policy requires max {policy.max_age_days} days",
                warnings,
            )

    # 5. Environment variable check
    missing = [var for var in policy.require_env if not os.environ.get(var)]
    if missing:
        _apply_enforcement(
            enf.mode_for("missing_env"),
            f"required environment variable not set: {missing[0]}",
            warnings,
        )

    # 6. SBOM check
    if policy.require_sbom and not has_sbom:
        _apply_enforcement(
            enf.mode_for("missing_sbom"),
            "package built without attestation slot — operator policy requires SBOM",
            warnings,
        )

    # 7. Trusted key check
    if policy.require_trusted_key and not key_trusted:
        _apply_enforcement(
            enf.mode_for("untrusted_key"),
            "operator policy requires a trusted signing key — package key is not in the trusted store",
            warnings,
        )

    return warnings


# 🌶️📦🔚
